aeron-archive/src/main/cpp_wrapper/client/archive/ArchiveContext.h

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2014-2024 Real Logic Limited.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef AERON_ARCHIVE_WRAPPER_CONTEXT_H
+#define AERON_ARCHIVE_WRAPPER_CONTEXT_H
+
+#include "AeronArchive.h"
+#include "CredentialsSupplier.h"
+#include "concurrent/YieldingIdleStrategy.h"
+
+#include "Context.h"
+
+namespace aeron { namespace archive { namespace client
+{
+
+class Context
+{
+public:
+    Context()
+    {
+        if (aeron_archive_context_init(&m_ctx) < 0)
+        {
+            ARCHIVE_MAP_ERRNO_TO_SOURCED_EXCEPTION_AND_THROW;
+        }
+
+        if (aeron_archive_context_set_credentials_supplier(
+            m_ctx,
+            encodedCredentialsFunc,
+            onChallengeFunc,
+            onFreeFunc,
+            (void *)this) < 0)
+        {
+            ARCHIVE_MAP_ERRNO_TO_SOURCED_EXCEPTION_AND_THROW;
+        }
+
+        if (aeron_archive_context_set_idle_strategy(
+            m_ctx,
+            idle,
+            (void *)this) < 0)
+        {
+            ARCHIVE_MAP_ERRNO_TO_SOURCED_EXCEPTION_AND_THROW;
+        }
+
+        this->idleStrategy();
+    }
+
+    ~Context()
+    {
+        aeron_archive_context_close(m_ctx);
+        m_ctx = nullptr;
+    }
+
+    std::string aeronDirectoryName()
+    {
+        return { aeron_archive_context_get_aeron_directory_name(m_ctx) };
+    }
+
+    void setAeronContext(const aeron::Context &ctx)
+    {
+        if (aeron_archive_context_set_aeron_context(m_ctx, ctx.context()) < 0)
+        {
+            ARCHIVE_MAP_ERRNO_TO_SOURCED_EXCEPTION_AND_THROW;
+        }
+    }
+
+    void credentialsSupplier(const CredentialsSupplier &credentialsSupplier)
+    {
+        m_credentialsSupplier.m_encodedCredentials = credentialsSupplier.m_encodedCredentials;
+        m_credentialsSupplier.m_onChallenge = credentialsSupplier.m_onChallenge;
+        m_credentialsSupplier.m_onFree = credentialsSupplier.m_onFree;
+    }
+
+    template<typename IdleStrategy = aeron::concurrent::YieldingIdleStrategy>
+    void idleStrategy()
+    {
+        // TODO I have no idea what is happening here...
+
+        IdleStrategy idleStrategy;
+
+        m_idleFunc = [&idleStrategy](int work_count){
```

Comment:
I'm trying to compensate for the fact that the various implemented idle strategies don't all inherit from a single class, so I can't set an idle strategy as a parameter.  I can do it via a template, however.  I think this works, but it sure does feel ugly.  I'm creating a lambda that wraps around the idle strategy that shows up because of the template.  I have no idea what sorts of naughty behaviours might result from this.  But... maybe it's ok?

Reply:
I think I figured this out.  The latest version of this function is simpler and makes more sense.